### ISSUE TITLE
Invalid matches

### ISSUE THREAD

**Issue by bgamari** [ChrisPenner/lens-regex-pcre#4]
I'm not sure if this is bug is in `lens-regex-pcre`, `pcre-heavy` (which I suspect), or `pcre-light` but this result is definitely wrong:
```
λ> import qualified Data.Text.IO as TIO
λ> import Control.Lens.Regex 
λ> import Control.Lens
λ> :set -XQuasiQuotes
λ> contents <- TIO.readFile "adjunctions-4.4.log" 
λ> let r = [rx|([\w\d ]+) \[([\w\d\.]+)\]: alloc=(\d+) time=(\d+\.\d+)|]
λ> putStrLn $ unlines $ map show $ take 10 $ drop 27 $ toListOf (regex r . groups) contents
["lifier [Da",".Functor.Contravariant.Rep]: a","120 tim","00\n***"]
["ialise [Da",".Functor.Contravariant.Rep]: a","84 tim","1\n***"]
["lifier [Da",".Functor.Contravariant.Rep]: a","968 tim","5\n***"]
["lifier [Da",".Functor.Contravariant.Rep]: a","752 tim","1\n***"]
["lifier [Da",".Functor.Contravariant.Rep]: a","752 tim","6\n***"]
["t inwards [Da",".Functor.Contravariant.Rep]: a"," tim","6\n***"]
["ed arity analysis [Da",".Functor.Contravariant.Rep]: a"," tim","9\n***"]
["lifier [Da",".Functor.Contravariant.Rep]: a","080 tim","6\n***"]
["nd analysis [Da",".Functor.Contravariant.Rep]: a","832 tim","9\n***"]
["er Wrapper binds [Da",".Functor.Contravariant.Rep]: a","04 tim","9\n***"]
```
These matches do not match the given regular expression.

**Comment by bgamari** [comment]
[adjunctions-4.4.log](https://github.com/ChrisPenner/lens-regex-pcre/files/3364811/adjunctions-4.4.log)

**Comment by bgamari** [comment]
As expected, the culprit is this line in the input:
```
219   ┊ ┊ because rule "Class op index" for ‘index’ might fire first↵                                                                                                                                                                 
```
This suggests that one of these libraries is not handling Unicode characters correctly.

**Comment by bgamari** [comment]
Applying `T.filter isAscii` to the input string is indeed enough to work around the issue in my case.

**Comment by ChrisPenner** [comment]
Indeed this is likely an issue with `pcre-light`;  I'd recommend making the appropriate issue there.

I'd also recommend trying compiling your regex with the `utf8` option; http://hackage.haskell.org/package/pcre-light-0.4.0.4/docs/Text-Regex-PCRE-Light.html

It notes:

> This option causes PCRE to regard both the pattern and the subject as strings of UTF-8 characters instead of single-byte character strings.

You may need to ensure your local PCRE is built with support for UTF-8 though.

Best of luck! Let me know if you find an appropriate fix 👍

**Comment by bgamari** [comment]
I think it would be best if this ticket remained open until a resolution is found. This is a rather serious correctness issue which your users are almost certain to hit. 

Sadly the `utf8` option won't help here as `Text`'s internal representation is UTF-16.

**Comment by ChrisPenner** [comment]
Turns out pcre-heavy returns all match offsets in bytes, but I was assuming it was Text characters.

I patched it by re-writing the implementation over ByteStrings instead. There's now also a `regexBS` combinator which should be more performant. `regex` should also be working now.

Re-open if this still gives you any trouble. Cheers!

Fixed in `v0.3.1.0`; http://hackage.haskell.org/package/lens-regex-pcre
